{{cookiecutter.package_name}}/src/{{cookiecutter.package_name}}/tools/make_artifacts.py

```diff
@@ -39,5 +39,154 @@ def build_artifacts(location: str, output_dir: str, append: bool, verbose: int):
         How noisy the logger should be.
 
     """
-    # Stubbed out pending further consideration of best location for this code.
-    pass
+    output_dir = Path(output_dir)
+
+    if location in project_globals.LOCATIONS:
+        path = Path(output_dir) / f'{sanitize_location(location)}.hdf'
+
+        if path.exists() and not append:
+            click.confirm(f"Existing artifact found for {location}. Do you want to delete and rebuild?",
+                          abort=True)
+            logger.info(f'Deleting artifact at {str(path)}.')
+            path.unlink()
+
+        build_single_location_artifact(path, location)
+
+    elif location == 'all':
+        # FIXME: could be more careful
+        existing_artifacts = set([item.stem for item in output_dir.iterdir()
+                                  if item.is_file() and item.suffix == '.hdf'])
+        locations = set([sanitize_location(loc) for loc in project_globals.LOCATIONS])
+        existing = locations.intersection(existing_artifacts)
+
+        if existing and not append:
+            click.confirm(f'Existing artifacts found for {existing}. Do you want to delete and rebuild?',
+                          abort=True)
+            for loc in existing:
+                path = output_dir / f'{loc}.hdf'
+                logger.info(f'Deleting artifact at {str(path)}.')
+                path.unlink()
+
+        build_all_artifacts(output_dir, verbose)
+
+    else:
+        raise ValueError(f'Location must be one of {project_globals.LOCATIONS} or the string "all". '
+                         f'You specified {location}.')
+
+
+def build_all_artifacts(output_dir: Path, verbose: int):
+    """Builds artifacts for all locations in parallel.
+
+    Parameters
+    ----------
+    output_dir
+        The directory where the artifacts will be built.
+    verbose
+        How noisy the logger should be.
+
+    Note
+    ----
+        This function should not be called directly.  It is intended to be
+        called by the :func:`build_artifacts` function located in the same
+        module.
+
+
+    """
+    from vivarium_cluster_tools.psimulate.utilities import get_drmaa
+    drmaa = get_drmaa()
+
+    jobs = {}
+    with drmaa.Session() as session:
```

Comment:
Let's look at other things we've had to do in parallel and see if we can't pull out a pattern:

- https://github.com/ihmeuw/vivarium_csu_ltbi/blob/master/src/vivarium_csu_ltbi/data/household_tb_scripts.py
- https://github.com/ihmeuw/vivarium_csu_ltbi/blob/master/src/vivarium_csu_ltbi/data/ltbi_incidence_scripts.py
- https://github.com/ihmeuw/vivarium_csu_hypertension_sdc/blob/master/src/vivarium_csu_hypertension_sdc/external_data/cli.py

Reply:
I think we could probably do something like 

```
def parallel_run(script_path: str, job_args: List[List[str]], cluster_args: NamedTuple, monitor: bool=True):
    """Run and monitor a script in parallel over the provided job args
    
    Parameters
    ----------------
    script:
        Path to an executable python script to be invoked with this python
        interpreter.
    job_args:
        A list whose elements are a list of arguments for each job to be run. 
    cluster_args:
        A named tuple with fields `project`, `queue`, `memory`, `cpu`, and `runtime`.

    """
    ...
```

Reply:
We could concat the script name with the args to make the job name.

Reply:
Monitoring seems like it might be hard, but maybe we can prototype this tool to only run in parallel over locations.

Reply:
Ticket created to examine generalizing existing parallel implementations